sdk/servicebus/Azure.Messaging.ServiceBus/src/Processor/SessionLifeCycleManager.cs

```diff
@@ -0,0 +1,251 @@
+                    _threadCount--;
+                    if (_threadCount == 0)
+                    {
+                        await CloseSession(cancellationToken).ConfigureAwait(false);
+                    }
+                }
+                else
+                {
+                    await CloseSession(cancellationToken).ConfigureAwait(false);
+                }
+            }
+            finally
+            {
+                if (releaseSemaphore)
+                {
+                    _semaphore.Release();
+                }
+            }
+        }
+
+        public async Task CloseSession(CancellationToken cancellationToken)
+        {
+            if (_receiver == null || _receiver.IsDisposed)
+            {
+                return;
+            }
+            try
+            {
+                if (_sessionCloseHandler != null)
+                {
+                    var args = new ProcessSessionEventArgs(_receiver, cancellationToken);
+                    await _sessionCloseHandler(args).ConfigureAwait(false);
+                }
+            }
+
+            catch (Exception exception)
+            {
+                await ProcessorUtils.RaiseExceptionReceived(
+                    _errorHandler,
+                    new ProcessErrorEventArgs(
+                        exception,
+                        ServiceBusErrorSource.CloseMessageSession,
+                        _fullyQualifiedNamespace,
+                        _entityPath))
+                    .ConfigureAwait(false);
+            }
+
+            finally
+            {
+                // Always at least attempt to dispose. If this fails, it won't be retried.
+                await _receiver.DisposeAsync().ConfigureAwait(false);
+                _receiver = null;
+            }
+        }
+
+        /// <summary>
+        ///
+        /// </summary>
+        /// <param name="receiver"></param>
+        /// <param name="cancellationTokenSource"></param>
+        /// <returns></returns>
+        private async Task RenewSessionLock(
+            ServiceBusSessionReceiver receiver,
+            CancellationTokenSource cancellationTokenSource)
+        {
+            cancellationTokenSource.CancelAfter(_maxAutoLockRenewal);
+            CancellationToken cancellationToken = cancellationTokenSource.Token;
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                try
+                {
+                    ServiceBusEventSource.Log.ProcessorRenewSessionLockStart(_identifier, receiver.SessionId);
+                    TimeSpan delay = ProcessorUtils.CalculateRenewDelay(receiver.SessionLockedUntil);
+
+                    // We're awaiting the task created by 'ContinueWith' to avoid awaiting the Delay task which may be canceled
+                    // by the renewLockCancellationToken. This way we prevent a TaskCanceledException.
+                    Task delayTask = await Task.Delay(delay, cancellationToken)
```

Comment:
You are still not avoiding TCE being thrown, why not suppress it here with try-catch?

Reply:
This was copied verbatim from Track 1 implementation. Are you saying not to use ContinueWith?

Reply:
Not for this reason...

Reply:
If the delay task is canceled we are breaking out of the loop. This would be different than awaiting the delay task and causing a TCE.